Feature/courses states task

irinache opened the pull request.

irinache committed:
Added basic bloc logic for course list

- Adde files course_list_bloc, course_list_events, course_list_state
- Changed CourseList to stateful widget

irinache committed:
Added uploading courses logic

- Updated course_list_bloc (added creation of new list of courseItems, hasMore constraints and updating of isLoading).
- Removed comment from course_list_events.
- Formatted code in course_list_state and changed default values of state.
- Added StreamBuilder to course_list.

idle-dt commented on lib/examples_from_courses/application/course_list_bloc/course_list_bloc.dart_bloc.dart:
```diff
@@ -0,0 +1,62 @@
+      final prevState = _stateController.value;
+      final items = newCourseItems(prevState.courseItems.length);
+
+      newItems.addAll(prevState.courseItems);
```
Показываю магию.
newItems.addAll([
...prevState.courseItems,
...items
])

idle-dt commented on lib/examples_from_courses/application/course_list_bloc/course_list_bloc.dart_bloc.dart:
```diff
@@ -0,0 +1,62 @@
+    }
+  }
+
+  List<String> newCourseItems(int lastIndex) {
```
Вы ведь не собираетесь вызывать этот метод где-то извне, верно? Значит это отличный претендент на то, чтобы стать приватным методом.

idle-dt replied:
А сможите сделать так, чтобы в этот метод передавалась просто текущая страничка, а не индекс последнего элемента?

idle-dt commented on lib/examples_from_courses/application/course_list_bloc/course_list_bloc.dart_bloc.dart:
```diff
@@ -0,0 +1,62 @@
+
+  List<String> newCourseItems(int lastIndex) {
+    final list = <String>[];
+    final n = min(_itemsPerPage, _finalCount - _currentPage * _itemsPerPage);
```
Нечитабельное название переменной.

idle-dt replied:
В остальном вы большие молодцы. Мне все нравится.

irinache committed:
Update course_list_bloc.dart_bloc.dart

- Updated adding items to new list.
- Changed _newCouseItems method: method converted to private, lastIndex removed from method parameters and calculated directly in method body.

idle-dt commented on lib/examples_from_courses/application/course_list_bloc/course_list_bloc.dart_bloc.dart:
```diff
@@ -0,0 +1,63 @@
+
+import 'course_list_events.dart';
+
+class CourseListBloc {
```
Код читается на ура. Кайф.

idle-dt commented on lib/examples_from_courses/presentation/pages/course_list/course_list.dart:
```diff
@@ -41,23 +44,49 @@ class CourseListPage extends StatelessWidget {
 }
 
-class _CourseList extends StatelessWidget {
+class _CourseList extends StatefulWidget {
```
Не вижу причин, по которым этот виджет должен оставаться Stateful, лучше пусть будет Stateless.

idle-dt commented on lib/examples_from_courses/presentation/pages/course_list/course_list.dart:
```diff
@@ -41,23 +44,49 @@ class CourseListPage extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) => StreamBuilder(
+        stream: _bloc.stateStream,
+        initialData: CourseListState.initial(),
```
В initialData нет не обходимости и так делать не надо. Вы в блоке используете BehaviorSubject. Он будет отдавать каждому новому подписчику последний эмит.

irinache committed:
Added small updates to course_list.dart

- Removed unused import.
- Changed CourseList to stateless widget.
- Removed initialData from StreamBuilder.
- itemCount calculations were completed and moved to funtion body.
